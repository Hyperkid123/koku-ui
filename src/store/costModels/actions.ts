import {
  CostModel,
  CostModelRequest,
  CostModels,
  deleteCostModel as apiDeleteCostModel,
  fetchCostModels as apiGetCostModels,
  updateCostModel as apiUpdateCostModel,
} from 'api/costModels';
import { AxiosError, AxiosResponse } from 'axios';
import * as H from 'history';
import { Dispatch } from 'react-redux';
import { createAsyncAction, createStandardAction } from 'typesafe-actions';

interface FilterQuery {
  currentFilterType?: string;
  currentFilterValue?: string;
}

export const updateFilterToolbar = createStandardAction(
  'fetch/costModels/filter'
)<FilterQuery>();

export const selectCostModel = createStandardAction('select/costModels')<
  CostModel
>();

export const resetCostModel = createStandardAction('reset/costModels')<void>();

interface DialogPayload {
  isOpen: boolean;
  name: string;
}

export const setCostModelDialog = createStandardAction(
  'display/costModels/dialog'
)<DialogPayload>();

export const {
  request: fetchCostModelsRequest,
  success: fetchCostModelsSuccess,
  failure: fetchCostModelsFailure,
} = createAsyncAction(
  'fetch/costModels/request',
  'fetch/costModels/success',
  'fetch/costModels/failure'
)<void, AxiosResponse<CostModels>, AxiosError>();

export const fetchCostModels = (query: string = '') => {
  return (dispatch: Dispatch) => {
    dispatch(fetchCostModelsRequest());

    return apiGetCostModels(query)
      .then(res => {
        dispatch(fetchCostModelsSuccess(res));
      })
      .catch(err => {
        dispatch(fetchCostModelsFailure(err));
      });
  };
};

export const {
  request: updateCostModelsRequest,
  success: updateCostModelsSuccess,
  failure: updateCostModelsFailure,
} = createAsyncAction(
  'update/costModels/request',
  'update/costModels/success',
  'update/costModels/failure'
)<void, AxiosResponse<CostModel>, AxiosError>();

export const updateCostModel = (
  uuid: string,
  request: CostModelRequest,
  dialog: string = null
) => {
  return (dispatch: Dispatch) => {
    dispatch(updateCostModelsRequest());

    return apiUpdateCostModel(uuid, request)
      .then(res => {
        dispatch(updateCostModelsSuccess(res));
        fetchCostModels()(dispatch);
        if (dialog !== null) {
          dispatch(setCostModelDialog({ name: dialog, isOpen: false }));
        }
      })
      .catch(err => {
        dispatch(updateCostModelsFailure(err));
      });
  };
};

export const {
  request: deleteCostModelsRequest,
  success: deleteCostModelsSuccess,
  failure: deleteCostModelsFailure,
} = createAsyncAction(
  'delete/costModels/request',
  'delete/costModels/success',
  'delete/costModels/failure'
)<void, void, AxiosError>();

export const deleteCostModel = (
  uuid: string,
  dialog: string = '',
  history: H.History = null
) => {
  return (dispatch: Dispatch) => {
    dispatch(deleteCostModelsRequest());

    return apiDeleteCostModel(uuid)
      .then(res => {
        dispatch(deleteCostModelsSuccess());
        dispatch(resetCostModel());
        fetchCostModels()(dispatch);
        if (dialog !== null) {
          if (dialog === 'deleteCostModel' && history) {
            history.push('/cost-models');
          }
          dispatch(setCostModelDialog({ name: dialog, isOpen: false }));
        }
      })
      .catch(err => {
        dispatch(deleteCostModelsFailure(err));
      });
  };
};
