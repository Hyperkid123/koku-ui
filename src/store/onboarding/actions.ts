import { addNotification } from '@red-hat-insights/insights-frontend-components/components/Notifications';
import { addProvider as apiCreateProvider } from 'api/providers';
import { Provider, ProviderRequest } from 'api/providers';
import { AxiosError } from 'axios';
import i18next from 'i18next';
import { Dispatch } from 'react-redux';
import { sourcesActions } from 'store/sourceSettings';
import {
  createAction,
  createAsyncAction,
  createStandardAction,
} from 'typesafe-actions';
import { Validator } from 'utils/validators';

export const updateType = createStandardAction('onboarding/update/type')<
  string,
  Validator
>();

export const updateName = createStandardAction('onboarding/update/name')<
  string,
  Validator
>();

export const updateClusterID = createStandardAction(
  'onboarding/update/clusterId'
)<string, Validator>();

export const updateS3BucketName = createStandardAction(
  'onboarding/update/s3BucketName'
)<string, Validator>();

export const updateArn = createStandardAction('onboarding/update/ARN')<
  string,
  Validator
>();

interface CheckPayload {
  item: string;
  value: boolean;
}

export const updateSourceKindCheckList = createStandardAction(
  'onboarding/update/sourceKindCheck'
)<CheckPayload>();

export const checkSourceKindCheckList = createAction(
  'onboarding/checkAll/sourceKindCheck'
);

export const openModal = createAction('onboarding/modal/open');

export const closeModal = createAction('onboarding/modal/close');

export const cancelOnboarding = createAction('onboarding/cancel');

export const {
  request: addSourceRequest,
  success: addSourceSuccess,
  failure: addSourceFailure,
} = createAsyncAction(
  'onboarding/source/add/request',
  'onboarding/source/add/success',
  'onboarding/source/add/failure'
)<void, Provider, AxiosError>();

export function addSource(request: ProviderRequest, sideEffect?: () => void) {
  return (dispatch: Dispatch) => {
    dispatch(addSourceRequest());
    return apiCreateProvider(request)
      .then(response => {
        dispatch(addSourceSuccess(response.data));
        sourcesActions.fetchSources()(dispatch);
        dispatch(
          addNotification({
            title: i18next.t('onboarding.notification.title', {
              name: request.name,
            }),
            description: i18next.t('onboarding.notification.description'),
            dismissable: true,
            variant: 'success',
          })
        );
        if (sideEffect !== null) {
          sideEffect();
        }
      })
      .catch(err => {
        dispatch(addSourceFailure(err));
      });
  };
}
