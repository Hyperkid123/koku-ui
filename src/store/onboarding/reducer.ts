import { AxiosError } from 'axios';
import { FetchStatus } from 'store/common';
import { ActionType, getType } from 'typesafe-actions';
import {
  addSourceFailure,
  addSourceRequest,
  addSourceSuccess,
  cancelOnboarding,
  checkKorekutaCheckList,
  checkSourceKindCheckList,
  closeModal,
  openModal,
  updateClusterID,
  updateKorekutaCheckList,
  updateName,
  updateSourceKindCheckList,
  updateType,
} from './actions';

export type Actions = ActionType<
  | typeof updateClusterID
  | typeof updateName
  | typeof updateType
  | typeof updateSourceKindCheckList
  | typeof updateKorekutaCheckList
  | typeof openModal
  | typeof closeModal
  | typeof cancelOnboarding
  | typeof checkSourceKindCheckList
  | typeof checkKorekutaCheckList
  | typeof addSourceRequest
  | typeof addSourceSuccess
  | typeof addSourceFailure
>;

type State = Readonly<{
  type: string;
  typeValid: boolean;
  typeDirty: boolean;
  name: string;
  nameValid: boolean;
  nameDirty: boolean;
  clusterId: string;
  clusterIdValid: boolean;
  clusterIdDirty: boolean;
  sourceKindChecks: object;
  korekutaChecks: object;
  isOpen: boolean;
  apiStatus: FetchStatus;
  apiErrors: AxiosError;
}>;

export const defaultState: State = {
  type: '',
  typeValid: true,
  typeDirty: false,
  name: '',
  nameValid: true,
  nameDirty: false,
  clusterId: '',
  clusterIdValid: true,
  clusterIdDirty: false,
  sourceKindChecks: {
    'check-ocp-version': false,
    'check-operator-metering': false,
    'check-insights-client': false,
    'check-ansible-epel': false,
    'check-oc': false,
  },
  korekutaChecks: {
    'check-ocp-api': false,
    'check-ocp-metering-operator-token-path': false,
    'check-ocp-operator-metering-namespace': false,
    'check-super-user-password': false,
  },
  isOpen: false,
  apiStatus: FetchStatus.none,
  apiErrors: null,
};

const areAllChecked = state => Object.keys(state).every(k => state[k]);

const setAll = (state, value) => {
  return Object.keys(state)
    .map(k => ({ [k]: value }))
    .reduce((acc, curr) => ({ ...acc, ...curr }));
};

export const stateKey = 'onboarding';

export function reducer(state: State = defaultState, action: Actions): State {
  switch (action.type) {
    case getType(updateType):
      return {
        ...state,
        type: action.payload,
        typeValid: action.meta(action.payload),
        typeDirty: true,
      };
    case getType(updateName):
      return {
        ...state,
        name: action.payload,
        nameValid: action.meta(action.payload),
        nameDirty: true,
      };
    case getType(updateClusterID):
      return {
        ...state,
        clusterId: action.payload,
        clusterIdValid: action.meta(action.payload),
        clusterIdDirty: true,
      };
    case getType(updateSourceKindCheckList):
      return {
        ...state,
        sourceKindChecks: {
          ...state.sourceKindChecks,
          [action.payload.item]: action.payload.value,
        },
      };
    case getType(updateKorekutaCheckList):
      return {
        ...state,
        korekutaChecks: {
          ...state.korekutaChecks,
          [action.payload.item]: action.payload.value,
        },
      };
    case getType(checkSourceKindCheckList):
      const skVal = areAllChecked(state.sourceKindChecks) ? false : true;
      return {
        ...state,
        sourceKindChecks: setAll(state.sourceKindChecks, skVal),
      };
    case getType(checkKorekutaCheckList):
      const kVal = areAllChecked(state.korekutaChecks) ? false : true;
      return {
        ...state,
        sourceKindChecks: setAll(state.korekutaChecks, kVal),
      };
    case getType(cancelOnboarding):
      return defaultState;
    case getType(openModal):
      return { ...state, isOpen: true };
    case getType(closeModal):
      return { ...state, isOpen: false };
    case getType(addSourceRequest):
      return {
        ...state,
        apiStatus: FetchStatus.inProgress,
        apiErrors: null,
      };
    case getType(addSourceFailure):
      return {
        ...state,
        apiStatus: FetchStatus.complete,
        apiErrors: action.payload,
      };
    case getType(addSourceSuccess):
      return {
        ...state,
        apiStatus: FetchStatus.complete,
        apiErrors: null,
      };
    default:
      return state;
  }
}
