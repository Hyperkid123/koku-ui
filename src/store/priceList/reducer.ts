import { Rates } from 'api/rates';
import { AxiosError } from 'axios';
import { FetchStatus } from 'store/common';
import { ActionType, getType } from 'typesafe-actions';
import {
  fetchPriceListFailure,
  fetchPriceListRequest,
  fetchPriceListSuccess,
} from './actions';

export const stateKey = 'priceList';

export type State = Readonly<{
  rates: Rates;
  error: AxiosError;
  status: FetchStatus;
}>;

export const defaultState: State = {
  rates: null,
  error: null,
  status: FetchStatus.none,
};

export type Action = ActionType<
  | typeof fetchPriceListRequest
  | typeof fetchPriceListSuccess
  | typeof fetchPriceListFailure
>;

export function reducer(state = defaultState, action: Action): State {
  switch (action.type) {
    case getType(fetchPriceListRequest):
      return {
        ...state,
        status: FetchStatus.inProgress,
      };
    case getType(fetchPriceListSuccess):
      return {
        ...state,
        rates: action.payload,
        error: null,
        status: FetchStatus.complete,
      };
    case getType(fetchPriceListFailure):
      return {
        ...state,
        error: action.payload,
        status: FetchStatus.complete,
      };
    default:
      return state;
  }
}
