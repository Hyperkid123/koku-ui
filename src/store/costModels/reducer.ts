import { CostModels } from 'api/costModels';
import { AxiosError } from 'axios';
import { FetchStatus } from 'store/common';
import { ActionType, getType } from 'typesafe-actions';
import {
  fetchCostModelsFailure,
  fetchCostModelsRequest,
  fetchCostModelsSuccess,
  updateFilterToolbar,
} from './actions';

export const stateKey = 'costModels';

export type CostModelsState = Readonly<{
  costModels: CostModels;
  error: AxiosError;
  status: FetchStatus;
  currentFilterType: string;
  currentFilterValue: string;
}>;

export const defaultState: CostModelsState = {
  costModels: null,
  error: null,
  status: FetchStatus.none,
  currentFilterType: 'name',
  currentFilterValue: '',
};

export type CostModelsAction = ActionType<
  | typeof fetchCostModelsFailure
  | typeof fetchCostModelsRequest
  | typeof fetchCostModelsSuccess
  | typeof updateFilterToolbar
>;

export const reducer = (
  state: CostModelsState = defaultState,
  action: CostModelsAction
): CostModelsState => {
  switch (action.type) {
    case getType(fetchCostModelsRequest):
      return {
        ...state,
        status: FetchStatus.inProgress,
      };
    case getType(fetchCostModelsSuccess):
      return {
        ...state,
        currentFilterValue: '',
        status: FetchStatus.complete,
        error: null,
        costModels: action.payload.data,
      };
    case getType(fetchCostModelsFailure):
      return {
        ...state,
        status: FetchStatus.complete,
        error: action.payload,
      };
    case getType(updateFilterToolbar):
      return {
        ...state,
        ...action.payload,
      };
    default:
      return state;
  }
};
