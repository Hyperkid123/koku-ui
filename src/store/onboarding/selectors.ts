import { RootState } from 'store/rootReducer';
import { stateKey } from './reducer';

export const selectOnboardingState = (state: RootState) => state[stateKey];

export const selectOnboardingName = (state: RootState) =>
  selectOnboardingState(state).name;

export const selectOnboardingConfirm = (state: RootState) =>
  selectOnboardingState(state).isConfirmShown;

export const selectOnboardingType = (state: RootState) =>
  selectOnboardingState(state).type;

export const selectOnboardingClusterID = (state: RootState) =>
  selectOnboardingState(state).clusterId;

export const selectOnboardingS3BucketName = (state: RootState) =>
  selectOnboardingState(state).s3BucketName;

export const selectOnboardingArn = (state: RootState) =>
  selectOnboardingState(state).arn;

export const selectOnboardingSourceKindChecked = (state: RootState) =>
  selectOnboardingState(state).sourceKindChecks;

export const selectOnboardingValidation = (state: RootState) => ({
  nameValid: selectOnboardingState(state).nameValid,
  typeValid: selectOnboardingState(state).typeValid,
  clusterIdValid: selectOnboardingState(state).clusterIdValid,
  s3BucketNameValid: selectOnboardingState(state).s3BucketNameValid,
  arnValid: selectOnboardingState(state).arnValid,
});

export const selectOnboardingDirty = (state: RootState) => {
  const obState = selectOnboardingState(state);
  return {
    name: obState.nameDirty,
    type: obState.typeDirty,
    clusterId: obState.clusterIdDirty,
    s3BucketName: obState.s3BucketNameDirty,
    arn: obState.arnDirty,
  };
};

export const selectOnboardingIsInvalid = (state: RootState) => {
  const data = selectOnboardingValidation(state);
  return Object.keys(data).some(k => data[k] === false);
};

export const selectOnboardingModal = (state: RootState) =>
  selectOnboardingState(state).isOpen;

export const selectApiErrors = (state: RootState) =>
  selectOnboardingState(state).apiErrors;

export const selectApiStatus = (state: RootState) =>
  selectOnboardingState(state).apiStatus;
