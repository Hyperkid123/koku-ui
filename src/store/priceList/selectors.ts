import { RootState } from 'store/rootReducer';
import { stateKey } from './reducer';

export const getRateTierTimeRange = (unit: string) => {
  const sep = unit.split('-');
  if (sep.length === 2) {
    return { unit: sep[0], range: sep[1] };
  }
  return { unit: null, period: null };
};

export const rateFlatter = rate => {
  return rate.tiered_rate.map((tier, ix) => {
    const time_range = getRateTierTimeRange(tier.usage.unit);
    return {
      provider_uuids: rate.provider_uuids,
      display: rate.metric.display_name,
      metric_type: rate.metric.name,
      index: ix,
      value: tier.value,
      value_unit: tier.unit,
      range_value: [tier.usage.usage_start, tier.usage.usage_end],
      range_unit: time_range.unit,
      period: time_range.range,
    };
  });
};

export const priceList = (state: RootState) => state[stateKey];

export const cachedRates = (state: RootState, providerUuid: string) => {
  if (priceList(state).rates && priceList(state).rates.get(providerUuid)) {
    return priceList(state).rates.get(providerUuid);
  }
  return null;
};

export const rates = (state: RootState, providerUuid: string) => {
  const cachedData = cachedRates(state, providerUuid);
  if (cachedData) {
    return cachedData.data;
  }
  return [];
};

export const ratesPerProvider = (state: RootState, providerUuid: string) =>
  rates(state, providerUuid) &&
  rates(state, providerUuid)
    .map(rateFlatter)
    .reduce((acc, rateArray) => {
      return [...acc, ...rateArray];
    }, [])
    .reduce((acc, rate) => {
      let next = { ...acc };
      rate.provider_uuids.forEach(uuid => {
        const prev = acc[uuid] || [];
        next = {
          ...next,
          [uuid]: [...prev, rate],
        };
      });
      return next;
    }, {});

export const status = (state: RootState, providerUuid: string) => {
  return priceList(state).status.get(providerUuid);
};

export const error = (state: RootState, providerUuid: string) =>
  priceList(state).error.get(providerUuid);
