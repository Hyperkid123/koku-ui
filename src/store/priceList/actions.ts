import { fetchRate } from 'api/rates';
import { Rates } from 'api/rates';
import { AxiosError } from 'axios';
import { Dispatch } from 'react-redux';
import { expirationMS, FetchStatus } from 'store/common';
import { RootState } from 'store/rootReducer';
import { createStandardAction } from 'typesafe-actions';

import { cachedRates, status } from './selectors';

interface Meta {
  providerUuid: string;
}

export const fetchPriceListRequest = createStandardAction('priceList/request')<Meta>();

export const fetchPriceListSuccess = createStandardAction('priceList/success')<Rates, Meta>();

export const fetchPriceListFailure = createStandardAction('priceList/failure')<AxiosError, Meta>();

function isExpired(state: RootState, meta: Meta) {
  const cachedData = cachedRates(state, meta.providerUuid);
  const reqStatus = status(state, meta.providerUuid);
  if (reqStatus === FetchStatus.inProgress) {
    return false;
  }

  if (!cachedData) {
    return true;
  }

  const now = Date.now();
  return now > cachedData.timeRequested + expirationMS;
}

export function fetchPriceList(providerUuid) {
  const meta = { providerUuid };
  return (dispatch: Dispatch, getState) => {
    if (!isExpired(getState(), meta)) {
      return;
    }
    dispatch(fetchPriceListRequest(meta));
    return fetchRate(providerUuid)
      .then(res => {
        dispatch(fetchPriceListSuccess(res.data, meta));
      })
      .catch(err => {
        dispatch(fetchPriceListFailure(err, meta));
      });
  };
}
