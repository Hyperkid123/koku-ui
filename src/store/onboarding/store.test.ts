import { createMockStoreCreator } from 'store/mockStore';
import * as actions from './actions';
import { reducer as obReducer, stateKey } from './reducer';
import * as selectors from './selectors';

const createObStore = createMockStoreCreator({
  [stateKey]: obReducer,
});

test('default state', async () => {
  const store = createObStore();
  expect(selectors.selectOnboardingState(store.getState())).toMatchSnapshot();
});

test('update a value alter it in store and set the validation', async () => {
  const validation = (value: string) => value !== 'bad_value';
  [
    {
      update: actions.updateType,
      selector: selectors.selectOnboardingType,
      validKey: 'typeValid',
    },
    {
      update: actions.updateClusterID,
      selector: selectors.selectOnboardingClusterID,
      validKey: 'clusterIdValid',
    },
    {
      update: actions.updateName,
      selector: selectors.selectOnboardingName,
      validKey: 'nameValid',
    },
  ].forEach(testCase => {
    const store = createObStore();
    store.dispatch(testCase.update('ok_value', validation));
    expect(testCase.selector(store.getState())).toBe('ok_value');
    expect(
      selectors.selectOnboardingValidation(store.getState())[testCase.validKey]
    ).toBe(true);

    store.dispatch(testCase.update('bad_value', validation));
    expect(testCase.selector(store.getState())).toBe('bad_value');
    expect(
      selectors.selectOnboardingValidation(store.getState())[testCase.validKey]
    ).toBe(false);
  });
});
