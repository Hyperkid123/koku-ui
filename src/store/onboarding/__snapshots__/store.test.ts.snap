// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default state 1`] = `
Object {
  "apiErrors": null,
  "apiStatus": 0,
  "clusterId": "",
  "clusterIdDirty": false,
  "clusterIdValid": true,
  "isOpen": false,
  "korekutaChecks": Object {
    "check-ocp-api": false,
    "check-ocp-metering-operator-token-path": false,
    "check-ocp-operator-metering-namespace": false,
    "check-super-user-password": false,
  },
  "name": "",
  "nameDirty": false,
  "nameValid": true,
  "sourceKindChecks": Object {
    "check-ansible-epel": false,
    "check-insights-client": false,
    "check-oc": false,
    "check-ocp-version": false,
    "check-operator-metering": false,
  },
  "type": "",
  "typeDirty": false,
  "typeValid": true,
}
`;
